File: class18/java/PES1802-DesignPatterns.java
/*
 * PES1802 - Padrões de Projeto GoF (Singleton, Factory Method, Observer, Strategy)
 * Aula 18: Princípios de Projeto e Padrões Arquiteturais
 * Projeto e Engenharia de Software - Prof. Giovane Barcelos
 *
 * Demonstra 4 padrões GoF com exemplos práticos:
 * - Singleton: gerenciador de configuração (instância única)
 * - Factory Method: criador de documentos (PDF, CSV, HTML)
 * - Observer: sistema de notificação (eventos de pedido)
 * - Strategy: calculadora de frete (PAC, SEDEX, transportadora)
 *
 * Compilar/executar:
 *   javac PES1802-DesignPatterns.java && java DesignPatterns
 */

import java.util.*;

// ============================================================
// 1. SINGLETON — Gerenciador de Configuração (instância única)
// ============================================================

class ConfigManager {
    private static ConfigManager instance;
    private Map<String, String> settings = new HashMap<>();

    private ConfigManager() {
        settings.put("app_name", "PES-EAD");
        settings.put("version", "1.0.0");
        settings.put("language", "pt-BR");
        settings.put("theme", "dark");
    }

    public static ConfigManager getInstance() {
        if (instance == null) {
            instance = new ConfigManager();
        }
        return instance;
    }

    public String get(String key) { return settings.getOrDefault(key, null); }

    public void set(String key, String value) { settings.put(key, value); }

    public Map<String, String> listAll() { return new HashMap<>(settings); }
}

// ============================================================
// 2. FACTORY METHOD — Criador de Documentos (PDF, CSV, HTML)
// ============================================================

interface Document {
    String render(List<Map<String, Object>> data);
}

class PdfDocument implements Document {
    public String render(List<Map<String, Object>> data) {
        StringBuilder sb = new StringBuilder();
        sb.append("=".repeat(40)).append("\n")
          .append("  RELATÓRIO PDF\n").append("=".repeat(40)).append("\n\n");
        for (var d : data) {
            sb.append(String.format("  %s | %-20s | R$ %.2f%n",
                d.get("id"), d.get("nome"), (Double) d.get("valor")));
        }
        sb.append("\n[PDF gerado com sucesso]");
        return sb.toString();
    }
}

class CsvDocument implements Document {
    public String render(List<Map<String, Object>> data) {
        StringBuilder sb = new StringBuilder("id,nome,valor\n");
        for (var d : data) {
            sb.append(String.format("%s,%s,%.2f%n",
                d.get("id"), d.get("nome"), (Double) d.get("valor")));
        }
        return sb.toString();
    }
}

class HtmlDocument implements Document {
    public String render(List<Map<String, Object>> data) {
        StringBuilder sb = new StringBuilder();
        sb.append("<html><body>\n  <h1>Relatório HTML</h1>\n  <table border='1'>\n");
        for (var d : data) {
            sb.append(String.format("    <tr><td>%s</td><td>%s</td><td>R$ %.2f</td></tr>%n",
                d.get("id"), d.get("nome"), (Double) d.get("valor")));
        }
        sb.append("  </table>\n</body></html>");
        return sb.toString();
    }
}

abstract class DocumentFactory {
    abstract Document createDocument();

    String export(List<Map<String, Object>> data) {
        return createDocument().render(data);
    }
}

class PdfExporter extends DocumentFactory {
    Document createDocument() { return new PdfDocument(); }
}

class CsvExporter extends DocumentFactory {
    Document createDocument() { return new CsvDocument(); }
}

class HtmlExporter extends DocumentFactory {
    Document createDocument() { return new HtmlDocument(); }
}

// ============================================================
// 3. OBSERVER — Sistema de Notificação de Pedidos
// ============================================================

interface Observer {
    void update(String event, Map<String, Object> payload);
}

abstract class Subject {
    protected List<Observer> observers = new ArrayList<>();

    void attach(Observer o) { observers.add(o); }
    void detach(Observer o) { observers.remove(o); }

    void notify(String event, Map<String, Object> payload) {
        for (Observer o : observers) o.update(event, payload);
    }
}

class PedidoService extends Subject {
    Map<String, Object> criarPedido(String cliente, double valor) {
        Map<String, Object> pedido = new HashMap<>();
        pedido.put("cliente", cliente);
        pedido.put("valor", valor);
        pedido.put("status", "CRIADO");
        notify("pedido.criado", pedido);
        return pedido;
    }

    void pagar(Map<String, Object> pedido) {
        pedido.put("status", "PAGO");
        notify("pedido.pago", pedido);
    }
}

class EmailNotifier implements Observer {
    public void update(String event, Map<String, Object> payload) {
        if ("pedido.criado".equals(event))
            System.out.printf("  📧 E-MAIL: Confirmação enviada para %s%n", payload.get("cliente"));
        else if ("pedido.pago".equals(event))
            System.out.printf("  📧 E-MAIL: Pagamento confirmado — R$ %.2f%n", payload.get("valor"));
    }
}

class LogNotifier implements Observer {
    public void update(String event, Map<String, Object> payload) {
        System.out.printf("  📋 LOG [%s] cliente=%s valor=R$ %.2f%n",
            event, payload.get("cliente"), payload.get("valor"));
    }
}

class SMSNotifier implements Observer {
    public void update(String event, Map<String, Object> payload) {
        if ("pedido.pago".equals(event))
            System.out.printf("  📱 SMS: Pagamento de R$ %.2f processado!%n", payload.get("valor"));
    }
}

// ============================================================
// 4. STRATEGY — Calculadora de Frete (PAC, SEDEX, Transportadora)
// ============================================================

interface FreteStrategy {
    double calcular(double pesoKg, double distanciaKm);
    String getNome();
}

class FretePAC implements FreteStrategy {
    public String getNome() { return "PAC (Correios)"; }
    public double calcular(double pesoKg, double distanciaKm) {
        return 12.0 + (pesoKg * 2.5) + (distanciaKm * 0.03);
    }
}

class FreteSEDEX implements FreteStrategy {
    public String getNome() { return "SEDEX (Correios)"; }
    public double calcular(double pesoKg, double distanciaKm) {
        return 18.0 + (pesoKg * 3.8) + (distanciaKm * 0.06);
    }
}

class FreteTransportadora implements FreteStrategy {
    public String getNome() { return "Transportadora Rápida"; }
    public double calcular(double pesoKg, double distanciaKm) {
        return 25.0 + (pesoKg * 1.8) + (distanciaKm * 0.04);
    }
}

class CalculadoraFrete {
    private FreteStrategy estrategia;

    CalculadoraFrete(FreteStrategy estrategia) { this.estrategia = estrategia; }
    void setEstrategia(FreteStrategy estrategia) { this.estrategia = estrategia; }

    String calcular(double pesoKg, double distanciaKm) {
        double valor = estrategia.calcular(pesoKg, distanciaKm);
        return String.format("Frete %s: R$ %.2f", estrategia.getNome(), valor);
    }
}

// ============================================================
// DEMONSTRAÇÃO
// ============================================================

class DesignPatterns {

    static void demoSingleton() {
        System.out.println("\n" + "─".repeat(55));
        System.out.println("  1. SINGLETON — Gerenciador de Configuração");
        System.out.println("─".repeat(55));
        ConfigManager cfg1 = ConfigManager.getInstance();
        ConfigManager cfg2 = ConfigManager.getInstance();
        System.out.printf("  Mesma instância? %b%n", cfg1 == cfg2);
        System.out.printf("  Config atual: %s%n", cfg1.listAll());
        cfg1.set("theme", "light");
        System.out.printf("  Após cfg1.set('theme','light'): cfg2.get('theme') = %s%n", cfg2.get("theme"));
    }

    static void demoFactoryMethod() {
        System.out.println("\n" + "─".repeat(55));
        System.out.println("  2. FACTORY METHOD — Exportação de Documentos");
        System.out.println("─".repeat(55));

        List<Map<String, Object>> dados = new ArrayList<>();
        dados.add(Map.of("id", 1, "nome", "Notebook", "valor", 4500.00));
        dados.add(Map.of("id", 2, "nome", "Mouse", "valor", 120.00));

        for (DocumentFactory f : new DocumentFactory[]{new PdfExporter(), new CsvExporter(), new HtmlExporter()}) {
            System.out.printf("%n  --- Exportando %s ---%n",
                f.getClass().getSimpleName().replace("Exporter", ""));
            System.out.println(f.export(dados));
        }
    }

    static void demoObserver() {
        System.out.println("\n" + "─".repeat(55));
        System.out.println("  3. OBSERVER — Notificação de Pedidos");
        System.out.println("─".repeat(55));

        PedidoService service = new PedidoService();
        service.attach(new EmailNotifier());
        service.attach(new LogNotifier());
        service.attach(new SMSNotifier());

        Map<String, Object> pedido = service.criarPedido("Maria Souza", 250.00);
        service.pagar(pedido);
    }

    static void demoStrategy() {
        System.out.println("\n" + "─".repeat(55));
        System.out.println("  4. STRATEGY — Cálculo de Frete");
        System.out.println("─".repeat(55));

        double peso = 5.0, distancia = 120.0;
        System.out.printf("  Pacote: %.1fkg, Distância: %.1fkm%n%n", peso, distancia);

        CalculadoraFrete calc = new CalculadoraFrete(new FretePAC());
        System.out.printf("  %s%n", calc.calcular(peso, distancia));

        calc.setEstrategia(new FreteSEDEX());
        System.out.printf("  %s%n", calc.calcular(peso, distancia));

        calc.setEstrategia(new FreteTransportadora());
        System.out.printf("  %s%n", calc.calcular(peso, distancia));
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(55));
        System.out.println("  GoF DESIGN PATTERNS — EXEMPLOS PRÁTICOS");
        System.out.println("=".repeat(55));

        demoSingleton();
        demoFactoryMethod();
        demoObserver();
        demoStrategy();

        System.out.println("\n" + "=".repeat(55));
        System.out.println("  ✓ Singleton: instância única (ConfigManager)");
        System.out.println("  ✓ Factory Method: criação delegada (Document -> PDF/CSV/HTML)");
        System.out.println("  ✓ Observer: notificação em cascata (Pedido -> Email/Log/SMS)");
        System.out.println("  ✓ Strategy: algoritmo intercambiável (Frete PAC/SEDEX/Transportadora)");
        System.out.println("=".repeat(55));
    }
}

File: class18/python/PES1802-DesignPatterns.py
"""
PES1802 - Padrões de Projeto GoF (Singleton, Factory Method, Observer, Strategy)
Aula 18: Princípios de Projeto e Padrões Arquiteturais
Projeto e Engenharia de Software - Prof. Giovane Barcelos

Demonstra 4 padrões GoF com exemplos práticos:
- Singleton: gerenciador de configuração (instância única)
- Factory Method: criador de documentos (PDF, CSV, HTML)
- Observer: sistema de notificação (eventos de pedido)
- Strategy: calculadora de frete (PAC, SEDEX, transportadora)
"""

from __future__ import annotations
from abc import ABC, abstractmethod
from typing import List

# ============================================================
# 1. SINGLETON — Gerenciador de Configuração (instância única)
# ============================================================

class ConfigManager:
    """Singleton: garante uma única instância de config global."""

    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance._settings = {
                "app_name": "PES-EAD",
                "version": "1.0.0",
                "language": "pt-BR",
                "theme": "dark",
            }
        return cls._instance

    def get(self, key: str, default=None):
        return self._settings.get(key, default)

    def set(self, key: str, value) -> None:
        self._settings[key] = value

    def list_all(self) -> dict:
        return dict(self._settings)


# ============================================================
# 2. FACTORY METHOD — Criador de Documentos (PDF, CSV, HTML)
# ============================================================

class Document(ABC):
    """Produto abstrato."""

    @abstractmethod
    def render(self, data: list[dict]) -> str:
        ...


class PdfDocument(Document):
    def render(self, data: list[dict]) -> str:
        header = "=" * 40 + "\n  RELATÓRIO PDF\n" + "=" * 40
        rows = "\n".join(f"  {d['id']} | {d['nome']:20s} | R$ {d['valor']:.2f}" for d in data)
        return f"{header}\n\n{rows}\n\n[PDF gerado com sucesso]"


class CsvDocument(Document):
    def render(self, data: list[dict]) -> str:
        rows = "\n".join(f"{d['id']},{d['nome']},{d['valor']:.2f}" for d in data)
        return f"id,nome,valor\n{rows}"


class HtmlDocument(Document):
    def render(self, data: list[dict]) -> str:
        rows = "\n".join(
            f"    <tr><td>{d['id']}</td><td>{d['nome']}</td><td>R$ {d['valor']:.2f}</td></tr>"
            for d in data
        )
        return (
            "<html><body>\n"
            "  <h1>Relatório HTML</h1>\n"
            "  <table border='1'>\n"
            f"{rows}\n"
            "  </table>\n"
            "</body></html>"
        )


class DocumentFactory(ABC):
    """Creator abstrato — Factory Method."""

    @abstractmethod
    def create_document(self) -> Document:
        ...

    def export(self, data: list[dict]) -> str:
        doc = self.create_document()
        return doc.render(data)


class PdfExporter(DocumentFactory):
    def create_document(self) -> Document:
        return PdfDocument()


class CsvExporter(DocumentFactory):
    def create_document(self) -> Document:
        return CsvDocument()


class HtmlExporter(DocumentFactory):
    def create_document(self) -> Document:
        return HtmlDocument()


# ============================================================
# 3. OBSERVER — Sistema de Notificação de Pedidos
# ============================================================

class Subject(ABC):
    """Subject (observável) — notifica observadores."""

    def __init__(self):
        self._observers: List[Observer] = []

    def attach(self, observer: "Observer") -> None:
        self._observers.append(observer)

    def detach(self, observer: "Observer") -> None:
        self._observers.remove(observer)

    def notify(self, event: str, payload: dict) -> None:
        for obs in self._observers:
            obs.update(event, payload)


class Observer(ABC):
    """Interface Observer."""

    @abstractmethod
    def update(self, event: str, payload: dict) -> None:
        ...


class PedidoService(Subject):
    """Serviço de pedidos — notifica observadores sobre eventos."""

    def criar_pedido(self, cliente: str, valor: float) -> dict:
        pedido = {"cliente": cliente, "valor": valor, "status": "CRIADO"}
        self.notify("pedido.criado", pedido)
        return pedido

    def pagar(self, pedido: dict) -> None:
        pedido["status"] = "PAGO"
        self.notify("pedido.pago", pedido)


class EmailNotifier(Observer):
    """Envia e-mail quando pedido é criado/pago."""

    def update(self, event: str, payload: dict) -> None:
        if event == "pedido.criado":
            print(f"  📧 E-MAIL: Confirmação enviada para {payload['cliente']}")
        elif event == "pedido.pago":
            print(f"  📧 E-MAIL: Pagamento confirmado — R$ {payload['valor']:.2f}")


class LogNotifier(Observer):
    """Registra log de todos os eventos."""

    def __init__(self):
        self._log: list[str] = []

    def update(self, event: str, payload: dict) -> None:
        entry = f"  📋 LOG [{event}] cliente={payload['cliente']} valor={payload['valor']}"
        self._log.append(entry)
        print(entry)


class SMSNotifier(Observer):
    """Envia SMS para eventos críticos."""

    def update(self, event: str, payload: dict) -> None:
        if event == "pedido.pago":
            print(f"  📱 SMS: Pagamento de R$ {payload['valor']:.2f} processado!")


# ============================================================
# 4. STRATEGY — Calculadora de Frete (PAC, SEDEX, Transportadora)
# ============================================================

class FreteStrategy(ABC):
    """Interface Strategy para cálculo de frete."""

    @abstractmethod
    def calcular(self, peso_kg: float, distancia_km: float) -> float:
        ...

    @property
    @abstractmethod
    def nome(self) -> str:
        ...


class FretePAC(FreteStrategy):
    """Correios PAC — mais barato, mais lento."""

    @property
    def nome(self) -> str:
        return "PAC (Correios)"

    def calcular(self, peso_kg: float, distancia_km: float) -> float:
        return 12.0 + (peso_kg * 2.5) + (distancia_km * 0.03)


class FreteSEDEX(FreteStrategy):
    """Correios SEDEX — mais rápido, mais caro."""

    @property
    def nome(self) -> str:
        return "SEDEX (Correios)"

    def calcular(self, peso_kg: float, distancia_km: float) -> float:
        return 18.0 + (peso_kg * 3.8) + (distancia_km * 0.06)


class FreteTransportadora(FreteStrategy):
    """Transportadora parceira — custo intermediário."""

    @property
    def nome(self) -> str:
        return "Transportadora Rápida"

    def calcular(self, peso_kg: float, distancia_km: float) -> float:
        return 25.0 + (peso_kg * 1.8) + (distancia_km * 0.04)


class CalculadoraFrete:
    """Contexto — usa uma estratégia de frete."""

    def __init__(self, estrategia: FreteStrategy):
        self._estrategia = estrategia

    def set_estrategia(self, estrategia: FreteStrategy) -> None:
        self._estrategia = estrategia

    def calcular(self, peso_kg: float, distancia_km: float) -> str:
        valor = self._estrategia.calcular(peso_kg, distancia_km)
        return f"Frete {self._estrategia.nome}: R$ {valor:.2f}"


# ============================================================
# DEMONSTRAÇÃO
# ============================================================

def demo_singleton():
    print("\n" + "─" * 55)
    print("  1. SINGLETON — Gerenciador de Configuração")
    print("─" * 55)
    cfg1 = ConfigManager()
    cfg2 = ConfigManager()
    print(f"  Mesma instância? {cfg1 is cfg2}")
    print(f"  Config atual: {cfg1.list_all()}")
    cfg1.set("theme", "light")
    print(f"  Após cfg1.set('theme','light'): cfg2.get('theme') = {cfg2.get('theme')}")


def demo_factory_method():
    print("\n" + "─" * 55)
    print("  2. FACTORY METHOD — Exportação de Documentos")
    print("─" * 55)
    dados = [
        {"id": 1, "nome": "Notebook", "valor": 4500.00},
        {"id": 2, "nome": "Mouse", "valor": 120.00},
    ]
    for exporter_class, label in [
        (PdfExporter, "PDF"),
        (CsvExporter, "CSV"),
        (HtmlExporter, "HTML"),
    ]:
        print(f"\n  --- Exportando {label} ---")
        print(exporter_class().export(dados))


def demo_observer():
    print("\n" + "─" * 55)
    print("  3. OBSERVER — Notificação de Pedidos")
    print("─" * 55)
    service = PedidoService()
    service.attach(EmailNotifier())
    service.attach(LogNotifier())
    service.attach(SMSNotifier())

    pedido = service.criar_pedido("Maria Souza", 250.00)
    service.pagar(pedido)


def demo_strategy():
    print("\n" + "─" * 55)
    print("  4. STRATEGY — Cálculo de Frete")
    print("─" * 55)
    peso, distancia = 5.0, 120.0
    print(f"  Pacote: {peso}kg, Distância: {distancia}km\n")

    calc = CalculadoraFrete(FretePAC())
    print(f"  {calc.calcular(peso, distancia)}")

    calc.set_estrategia(FreteSEDEX())
    print(f"  {calc.calcular(peso, distancia)}")

    calc.set_estrategia(FreteTransportadora())
    print(f"  {calc.calcular(peso, distancia)}")


if __name__ == "__main__":
    print("=" * 55)
    print("  GoF DESIGN PATTERNS — EXEMPLOS PRÁTICOS")
    print("=" * 55)
    demo_singleton()
    demo_factory_method()
    demo_observer()
    demo_strategy()

    print("\n" + "=" * 55)
    print("  ✓ Singleton: instância única (ConfigManager)")
    print("  ✓ Factory Method: criação delegada (Document -> PDF/CSV/HTML)")
    print("  ✓ Observer: notificação em cascata (Pedido -> Email/Log/SMS)")
    print("  ✓ Strategy: algoritmo intercambiável (Frete PAC/SEDEX/Transportadora)")
    print("=" * 55)
